ompi/mca/osc/ucx/osc_ucx_component.c

```diff
@@ -0,0 +1,565 @@
+    } else { /* different disp_unit sizes, allocate O(p) space to store them */
+        module->disp_unit = -1;
+        module->disp_units = calloc(comm_size, sizeof(int));
+        if (module->disp_units == NULL) {
+            goto error;
+        }
+
+        ret = module->comm->c_coll->coll_allgather(&disp_unit, 1, MPI_INT,
+                                                   module->disp_units, 1, MPI_INT,
+                                                   module->comm,
+                                                   module->comm->c_coll->coll_allgather_module);
+        if (OMPI_SUCCESS != ret) {
+            goto error;
+        }
+    }
+
+    /* exchange endpoints if necessary */
+    for (i = 0; i < comm_size; i++) {
+        if (OSC_UCX_GET_EP(module->comm, i) == NULL) {
+            collect_eps = true;
+            break;
+        }
+    }
+
+    if (collect_eps) {
+        int disps[comm_size];
+
+        status = ucp_worker_get_address(mca_osc_ucx_component.ucp_worker,
+                                        &my_addr, &my_addr_len);
+        if (status != UCS_OK) {
+            goto error;
+        }
+
+        ret = allgather_len_and_info(my_addr, (int)my_addr_len,
+                                     &recv_buf, disps, module->comm);
+        if (ret != OMPI_SUCCESS) {
+            goto error;
+        }
+
+        for (i = 0; i < comm_size; i++) {
+            if (OSC_UCX_GET_EP(module->comm, i) == NULL) {
+                ucp_ep_params_t ep_params;
+                ucp_ep_h ep;
+                memset(&ep_params, 0, sizeof(ucp_ep_params_t));
+                ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
+                ep_params.address = (ucp_address_t *)&(recv_buf[disps[i]]);
+                status = ucp_ep_create(mca_osc_ucx_component.ucp_worker, &ep_params, &ep);
+                if (status != UCS_OK) {
+                    goto error;
+                }
+
+                ompi_comm_peer_lookup(module->comm, i)->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_UCX] = ep;
+            }
+        }
+
+        ucp_worker_release_address(mca_osc_ucx_component.ucp_worker, my_addr);
+        my_addr = NULL;
+        free(recv_buf);
+        recv_buf = NULL;
+    }
+
+    /* initialize UCP win_info array */
+    module->win_info_array = calloc(comm_size, sizeof(ompi_osc_ucx_win_info_t));
+    if (module->win_info_array == NULL) {
+        goto error;
+    }
+
+    ret = mem_map_and_exchange_win_info(base, size, &(module->memh), module,
+                                        module->win_info_array, flavor);
+    if (ret != OMPI_SUCCESS) {
+        goto error;
+    }
+
+    /* initialize UCP state_info array */
+    module->state_info_array = calloc(comm_size, sizeof(ompi_osc_ucx_win_info_t));
+    if (module->state_info_array == NULL) {
+        goto error;
+    }
+
+    state_base = (void *)&(module->state);
+    ret = mem_map_and_exchange_win_info(&state_base, sizeof(ompi_osc_ucx_state_t),
```

Comment:
maybe it's better to collect state_info and win_info in one collective call somehow?